Feat/172 add h2 demand

lisazeyen opened the pull request:
Closes #173 and #172. 

## Changes proposed in this Pull Request
Adding exogenous hydrogen demand to the network and modelling the NT scenario with only one hydrogen zone.

## Tasks

- [x] move definition of Z1 and Z2 buses to spatial function
- [x] make H2 topology a switch in the config instead of dependency on tnydp scenario name

## Workflow
### Add configurable hydrogen topology for TYNDP scenarios
This PR introduces a new configuration setting `h2_zones_tyndp` (bool) that controls whether the hydrogen topology is modeled as a single zone or split into two separate zones (Z1 and Z2). This flexibility was added to accommodate potential changes in TYNDP methodology across different cycles. Currently, the configuration works with the available input data: DE/GA scenarios use a split topology (`true`), while NT uses a single zone (`false`).

### Implementation details:
In define_spatial, the function now creates Z1 and Z2 hydrogen buses based on the selected topology configuration. When modeling a single zone, the Z1 bus is defined as an empty index. The naming convention for buses and assets follows the previous approach: the split topology uses the suffix `"H2 Z1/Z2"`, while the single-zone configuration simply uses `"H2"`.

In `prepare_sector_network`, a new function `add_h2_demand_tyndp` has been added to process and attach exogenous hydrogen demand data to the network.

### Asset attachment
| Asset Category | Asset Type | Split Topology (Z1/Z2) | Single Zone |
|---|---|---|---|
| **H2 Production** | Electrolysis | Z1, Z2 | H2 |
| | SMR (optional) | Z1 | H2 |
| | SMR CC (optional) | Z1 | H2 |
| | ATR (optional)| Z1 | H2 |
| **H2 Reconversion** | Fuel cells (optional) | Z2 | H2 |
| | H2 turbines (optional) | Z2 | H2 |
| | Methanation (optional) | Z2 | H2 |
| **H2 Storage** | H2 tanks | Z1 | — |
| | Salt caverns | Z2 | H2 |

### Results for H2 Balance and Oil Balance
<img width="3548" height="1760" alt="oil_energy_balance_comparison" src="https://github.com/user-attachments/assets/51e125aa-5f88-4e9a-9758-f37738c3df77" />
<img width="3556" height="1760" alt="H2_energy_balance_comparison" src="https://github.com/user-attachments/assets/4b395eb4-b4c4-4d95-8ad5-c1ba0f378206" />

## Open issues
- currently the industry sector is still considered for the NT scenarios, resulting in an additional demand for hydrogen to produce synthetic fuels (Fischer-Tropsch). This issue will be closed by PR #220 
 - for the benchmarking the aggregated exogenous hydrogen demand needs to be split for NT into industry/transport/...
 - add H2 tanks also in NT scenario?
 
## Notes


## Checklist

<!-- Remove what doesn't apply. -->

- [x] I tested my contribution locally and it works as intended.
- [x] Code and workflow changes are sufficiently documented.
- [x] Changed dependencies are added to `envs/environment.yaml`.
- [x] Changes in configuration options are added in `config/config.default.yaml`.
- [x] Changes in configuration options are documented in `doc/configtables/*.csv`.
- [x] Changes in configuration options are added in `config/test/*.yaml`.
- [x] Open-TYNDP SPDX license header added to all touched files.
- [x] Sources of newly added data are documented in `doc/data_sources.rst`.
- [x] New rules are documented in the appropriate `doc/*.rst` files.
- [x] A release note `doc/release_notes.rst` is added.
- [x] Major features are documented with up-to-date information in `README` and `doc/index.rst`.
- [x] For any new/added Python scripts: Module docstring is added describing what the script does, its input data, and its output data.

daniel-rdt commented on scripts/prepare_sector_network.py:
```diff
@@ -1947,16 +1980,15 @@ def add_h2_production_tyndp(n, nodes, buses_h2_z1, costs, options={}):
         v if isinstance(v, bool) else any(v.values())
```
If I'm not misreading this, ATR is also optional. In the PR description it is not marked as optional however.

daniel-rdt commented on scripts/prepare_sector_network.py:
```diff
@@ -2905,6 +2973,8 @@ def add_gas_and_h2_infrastructure(
     gas_input_nodes,
     spatial,
     options,
+    tyndp_scenario,
```
Please, add to the docstrings

daniel-rdt commented on scripts/prepare_sector_network.py:
```diff
@@ -200,17 +202,37 @@ def define_spatial(nodes, options, offshore_buses_fn=None, buses_h2_file=None):
-            vars(spatial.h2_tyndp),
-            index=pd.Index((buses_h2.index + " Z1").append(buses_h2.index + " Z2")),
-        )
+
```
Nice how you managed to implement the spatial split!

daniel-rdt commented on scripts/prepare_sector_network.py:
```diff
@@ -2401,12 +2434,10 @@ def add_h2_reconversion_tyndp(n, spatial, nodes, buses_h2_z2, costs, options=Non
             "Link",
-            nodes.index + " H2 Z2 h2-fuel-cell",
-            bus0=nodes.country.values + " H2 Z2",
+            nodes.index + f" {suffix} Fuel Cell",
```
I think this one got lost when merging the master. It should remain as h2-fuel-cell
```suggestion
            nodes.index + f" {suffix} h2-fuel-cell",
```

daniel-rdt commented on scripts/prepare_sector_network.py:
```diff
@@ -6388,10 +6468,11 @@ def add_industry(
                 )
 
     # TODO Link properly Industry to H2 topology
+    buses_h2 = spatial.buses_h2_z2 if options["h2_topology_tyndp"] else nodes_ind_h2
```
I just stumbled over this as I was reviewing the changes and maybe we want to change the variable slightly, so it keeps with the default PyPSA-Eur naming convention when we are not modelling the tyndp topology. At the moment, for the case of `h2_topology_tyndp=False` the FT Links will be called eg. `DE00 H2 Fischer-Tropsch` instead of before when it was called only `DE00 Fischer-Tropsch`.
But maybe it is also just fine as is.

daniel-rdt commented on scripts/prepare_sector_network.py:
```diff
@@ -2766,6 +2809,44 @@ def add_h2_topology_tyndp(
+        bus=demand.columns,
+        carrier="H2",
+        p_set=demand,
+    )
```
Addressed by https://github.com/open-energy-transition/open-tyndp/pull/230/commits/82fc1b75a92ef391a7453daf6cf5d1479964aeae

lisazeyen commented:
> Hi @lisazeyen, Thomas and I ran into one naming issue with the added H2 load. We either need to change or remove the nice name for H2 if we keep H2 as the carrier for the load. Alternatively, we could change the carrier for the load to smth like `H2 exogenous demand` and add the carrier with a dedicated color. Any thoughts?
> 
> Otherwise, thank you for addressing the comments, the PR works great and looks almost ready now, only two more small comments.

Thanks @daniel-rdt for the review and carrier renaming! I've addressed all your comments. Let me know if the FT link renaming works for you.

lisazeyen commented:
> > Thanks @daniel-rdt for the review and carrier renaming! I've addressed all your comments. Let me know if the FT link renaming works for you.
> 
> Great, the renamed FT links should work well, thanks @lisazeyen :)
> 
> Regarding the edge case of no defined TYNDP scenario, I was wondering if we should already make the demand input optional in `prepare_sector_network`, so that the `build_tyndp_h2_demand` rule will not have to be executed at all. This would mean that in the default config we define:
> 
> ```
> tyndp_scenario: false  # 'NT', 'DE', 'GA' or false
> ```
> 
> and then for the input of `prepare_sector_network` we could use branch:
> 
> ```
> h2_demand=branch(
>             config_provider("tyndp_scenario"),
>             resources("h2_demand_tyndp_{planning_horizons}.csv"),
>             [],
>         ),
> ```
> 
> Likewise, we could do the same for the gas_demand input:
> 
> ```
> gas_demand=branch(
>             config_provider("tyndp_scenario"),
>             resources("gas_demand_tyndp_{planning_horizons}.csv"),
>             [],
>         ),
> ```
> 
> I think that should work and clean up the dag a little for non-TYNDP runs

I like the idea! Let's do it in a separate PR!

daniel-rdt reviewed (approved):
Thank you @lisazeyen. Everything looks great now! I am not sure why the whole sector.csv file is marked as diff but I checked the docs and it is looking like it should!

Let's wait for the Biomass PR first and then LGTM 🎊
